File: src/Warehouse.Core.Functional/Pipelines/Trading/TradingContext.fs
namespace Warehouse.Core.Functional.Pipelines.Trading

open System
open Warehouse.Core.Functional.Markets.Contracts
open Warehouse.Core.Functional.Markets.Domain
open Warehouse.Core.Functional.Pipelines.Contracts

type TradingContext() =
    interface IPipelineContext with
        member this.ExecutionId = this.ExecutionId
        member this.StartedAt = this.StartedAt
        member this.IsCancelled = this.IsCancelled

        member this.IsCancelled
            with set value = this.IsCancelled <- value

    member val PipelineId = 0 with get, set
    member val CurrentMarketData: MarketData option = None with get, set
    member val MarketType = Unchecked.defaultof<MarketType> with get, set
    member val Symbol = "" with get, set
    member val BuyPrice: decimal option = None with get, set
    member val Quantity: decimal option = None with get, set
    member val Action = TradingAction.None with get, set
    member val ActiveOrderId: int64 option = None with get, set
    member val CurrentPrice = 0m with get, set
    member val ExecutionId = Guid.CreateVersion7() with get
    member val StartedAt = DateTime.UtcNow with get
    member val IsCancelled = false with get, set

File: src/Warehouse.Core.Old/Pipelines/Trading/TradingContext.cs
using Warehouse.Core.Old.Functional.Markets.Contracts;
using Warehouse.Core.Old.Functional.Markets.Domain;
using Warehouse.Core.Old.Pipelines.Core;

namespace Warehouse.Core.Old.Pipelines.Trading;

public class TradingContext : IPipelineContext
{
    public required int PipelineId { get; init; }

    public MarketData? CurrentMarketData { get; set; }

    public MarketType MarketType { get; init; }

    public required string Symbol { get; init; }

    public decimal? BuyPrice { get; set; }

    public decimal? Quantity { get; set; }

    public TradingAction Action { get; set; } = TradingAction.None;

    public long? ActiveOrderId { get; set; }

    public decimal CurrentPrice { get; set; }

    public Guid ExecutionId { get; } = Guid.CreateVersion7();

    public DateTime StartedAt { get; } = DateTime.UtcNow;

    public bool IsCancelled { get; set; }
}
